src/easylink/step.py

```diff
@@ -730,3 +735,114 @@ def _update_slot_mappings(self, num_repeats) -> dict[str, list[SlotMapping]]:
             for slot in self.output_slots
         ]
         return {"input": input_mappings, "output": output_mappings}
+
+
+class ChoiceStep(Step):
+    """A ChoiceStep allows a user to select a single path from a set of possible paths."""
+
+    def __init__(
+        self,
+        step_name: str,
+        input_slots: Iterable[InputSlot],
+        output_slots: Iterable[OutputSlot],
+        choices: dict[
+            str, dict[str, list[Step | EdgeParams | InputSlotMapping | OutputSlotMapping]]
+        ],
+    ) -> None:
+        super().__init__(
+            step_name,
+            input_slots=input_slots,
+            output_slots=output_slots,
+        )
+        self.choices = choices
+
+    def validate_step(
+        self,
+        step_config: LayeredConfigTree,
+        combined_implementations: LayeredConfigTree,
+        input_data_config: LayeredConfigTree,
+    ) -> dict[str, list[str]]:
+        """Validates the ChoiceStep.
+
+        Notes
+        -----
+        We update the step graph and slot mappings here in validation as opposed to
+        in set_configuration_state (as is done in TemplatedSteps) because ChoiceStep
+        validation happens prior to set_configuration_state and actually requires
+        the step graph and mappings.
+
+        We do not attempt to validate the subgraph here if the 'type' key is unable
+        to be validated.
+        """
+
+        chosen_type = step_config.get("type")
+        # Handle problems with the 'type' key
+        if not chosen_type:
+            return {f"step {self.name}": ["The step requires a 'type' key."]}
+        if chosen_type not in self.choices:
+            return {
+                f"step {self.name}": [
+                    f"'{step_config['type']}' is not a supported 'type'. Valid choices are: {list(self.choices)}."
+                ]
+            }
+        # Handle type-subgraph inconsistencies
+        subgraph = self.choices[chosen_type]
+        chosen_step_config = LayeredConfigTree(
+            {key: value for key, value in step_config.items() if key != "type"}
+        )
+        allowable_steps = [node.name for node in subgraph["nodes"]]
+        if set(allowable_steps) != set(chosen_step_config):
+            return {
+                f"step {self.name}": [
+                    f"Invalid configuration for '{chosen_type}' type. Valid steps are {allowable_steps}."
+                ]
+            }
+
+        # Handle the actual chosen step_config
+        self.step_graph = self._update_step_graph(subgraph)
```

Comment:
I think instead you can re-use `Step._get_step_graph`, i.e. `self.step_graph = self._get_step_graph(subgraph['nodes'],subgraph['edges'])`